illumos-utils/src/running_zone.rs

```diff
@@ -137,9 +140,102 @@ pub enum GetZoneError {
     },
 }
 
+// Helper module for setting up and running `zone_enter()` for subprocesses run
+// inside a non-global zone.
+#[cfg(target_os = "illumos")]
+mod zenter {
+    use libc::zoneid_t;
+    use std::ffi::c_int;
+    use std::ffi::c_uint;
+    use std::ffi::CStr;
+
+    #[link(name = "contract")]
+    extern "C" {
+        fn ct_tmpl_set_critical(fd: c_int, events: c_uint) -> c_int;
+        fn ct_tmpl_set_informative(fd: c_int, events: c_uint) -> c_int;
+        fn ct_pr_tmpl_set_fatal(fd: c_int, events: c_uint) -> c_int;
+        fn ct_pr_tmpl_set_param(fd: c_int, params: c_uint) -> c_int;
+        fn ct_tmpl_activate(fd: c_int) -> c_int;
+        fn ct_tmpl_clear(fd: c_int) -> c_int;
+    }
+
+    #[link(name = "c")]
+    extern "C" {
+        pub fn zone_enter(zid: zoneid_t) -> c_int;
+    }
+
+    // A Rust wrapper around the process contract template.
+    #[derive(Debug)]
+    pub struct Template {
+        fd: c_int,
+    }
+
+    impl Drop for Template {
+        fn drop(&mut self) {
+            if unsafe { libc::close(self.fd) } != 0 {
+                let e = std::io::Error::last_os_error();
+                eprintln!("failed to close ctfs template file: {e}");
```

Comment:
It probably will, yes. I'll look into getting a logger down here, though it might take a bit more work since I don't believe the `RunningZone` object has one either.

Reply:
There's no straightforward way to get a logger here, since we have to go quite a ways up the object tree to get to one that has a logger. Bummer, but we'll live with it for now.

Reply:
I'm very tempted to propose removing the `eprintln!()` altogether then. :-/ I worry this will get intermingled with slog output is a weird way. Not logging errors on close is not great but does have precedence: normal `File`s do the same thing.

Reply:
Sure that seems OK to me too. I'm pretty agnostic. We also mix an `eprintln!()` call with `slog` output when we fail to clean up other resources, like VNICs. So there's even precedent within sled-agent! But it's still not the best. Will remove for now.

Reply:
Removed in 58c78606c